### PR TITLE
Implementation of Attendance system

### PR DESCRIPTION
**Pull request by Asheraf**
[//]: # (**********************************)
[//]: # (** Fill in the following fields **)
[//]: # (**********************************)

[//]: # (Note: Lines beginning with syntax such as this one, are comments and will not be visible in your report!)

### Pull Request Prelude

[//]: # (Thank you for working on improving Hercules!)

[//]: # (Please complete these steps and check the following boxes by putting an `x` inside the brackets _before_ filing your Pull Request.)

- [x] I have followed [proper Hercules code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR will be closed if the above-mentioned criteria are not fulfilled.

This PR Implements the support for the new attendance UI that got introduced to the clients as of 2018-03-07bRagexeRE .

[//]: # (Insert checklist here)
[//]: # (Syntax: - [ ] Checkbox)

[//]: # (**NOTE** Enable the setting "[√] Allow edits from maintainers." when creating your pull request if you have not already enabled it.)

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

### CODE REVIEW

**Comment by HerculesWSAPI** [comment]
This change is [Reviewable](https://reviewable.io/reviews/herculesws/hercules/1990)

File: sql-files/upgrades/2018-03-10--04-06.sql [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+-- GNU General Public License for more details.
+--
+-- You should have received a copy of the GNU General Public License
+-- along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+ALTER TABLE `char` ADD COLUMN `attendance_count` TINYINT(3) UNSIGNED NOT NULL DEFAULT '0' AFTER `hotkey_rowshift`;
+ALTER TABLE `char` ADD COLUMN `attendance_timer` BIGINT(20) NULL DEFAULT '0' AFTER `attendance_count`;
+
+INSERT INTO `sql_updates` (`timestamp`, `ignored`) VALUES (1520654809 , 'No');
```
Comment (4144) [inline sql-files/upgrades/2018-03-10--04-06.sql]: missing update in index.txt

File: src/map/clif.c [hunk 2]
```diff
@@ -20186,6 +20186,146 @@ void clif_skill_scale(struct block_list *bl, int src_id, int x, int y, uint16 sk
...
+		clif->attendancedb_libconfig_sub(it, i, config_filename);
+	}
+
+	libconfig->destroy(&attendance_conf);
+	ShowStatus("Done reading '"CL_WHITE"%d"CL_RESET"' entries in '"CL_WHITE"%s"CL_RESET"'.\n", i, config_filename);
+	return true;
+}
+
+bool clif_attendancedb_libconfig_sub(struct config_setting_t *it, int n, const char *source) {
+	struct config_setting_t *t = NULL;
```
Comment (4144) [inline src/map/clif.c]: here need check it, source for null with nullpo
and check n for valid index with assertpo

File: src/map/clif.h [hunk 3]
```diff
@@ -574,6 +576,20 @@ enum CZ_CONFIG {
...
 };
+/**
+* Client UI types
+**/
+enum ui_types {
+	BANK_UI = 0,
+	STYLIST_UI,
+	CAPTCHA_UI,
+	MACRO_UI,
+	// unknown,
```
Comment (4144) [inline src/map/clif.h]: here unused

File: src/map/clif.c [hunk 4]
```diff
@@ -20482,6 +20622,13 @@ void do_final_clif(void)
 			aFree(clif->rd.qty[i]);
 	}
 
+	for (i = 0; i < MAX_ATTENDANCE_DAYS; i++) {
+		if (clif->attendance_data.nameid[i])
+			aFree(clif->attendance_data.nameid[i]);
```
Comment (4144) [inline src/map/clif.c]: you cant remove int values.
if need clean id's set array fields to 0.

same for second aFree.

File: src/map/clif.c [hunk 5]
```diff
@@ -20186,6 +20186,162 @@ void clif_skill_scale(struct block_list *bl, int src_id, int x, int y, uint16 sk
...
+
+void clif_parse_attendance_reward_request(int fd, struct map_session_data *sd) __attribute__((nonnull(2)));
+void clif_parse_attendance_reward_request(int fd, struct map_session_data *sd)
+{
+#if defined(PACKETVER_RE) && PACKETVER >= 20180307
+
+	if (battle_config.feature_enable_attendance_system != 1)
+		return;
+
+	nullpo_retv(sd);
```
Comment (4144) [inline src/map/clif.c]: here sd check useless, because __attribute__((nonnull(2)) said it must be non null

File: src/map/clif.c [hunk 6]
```diff
@@ -20186,6 +20186,165 @@ void clif_skill_scale(struct block_list *bl, int src_id, int x, int y, uint16 sk
...
+}
+
+bool clif_attendancedb_libconfig_sub(struct config_setting_t *it, int n, const char *source)
+{
+	struct config_setting_t *t = NULL;
+	struct item_data *data = NULL;
+	int i32 = 0;
+	const char *str = NULL;
+
+	nullpo_ret(source);
```
Comment (4144) [inline src/map/clif.c]: missing nullpo for it

File: src/map/clif.c [hunk 7]
```diff
@@ -20222,6 +20222,169 @@ void clif_skill_scale(struct block_list *bl, int src_id, int x, int y, uint16 sk
...
+
+
+	struct PACKET_ZC_OPEN_UI p;
+	int claimed = 0;
+
+	nullpo_retv(sd);
+
+	p.PacketType = 0xAE2;
+	switch (UIType) {
+	case 5: // receive 5 for ATTENDANCE_UI
```
Comment (Lemongrass3110) [inline src/map/clif.c]: Your enum ui_types seems to be not used and does not match with your code here.

**Comment by Asheraf** [comment]
Here 2 types a zc and cz and they are different, client send 5 for the
request but need to receive 7 to open.

On Mar 11, 2018 11:30 PM, "Lemongrass3110" <notifications@github.com> wrote:

*@Lemongrass3110* commented on this pull request.
------------------------------

In src/map/clif.c
<https://github.com/HerculesWS/Hercules/pull/1990#discussion_r173674609>:

> +	clif->open_ui(sd, p->UIType);
+}
+
+void clif_open_ui(struct map_session_data *sd, int8 UIType)
+{
+#if defined(PACKETVER_RE) && PACKETVER >= 20180307
+
+
+	struct PACKET_ZC_OPEN_UI p;
+	int claimed = 0;
+
+	nullpo_retv(sd);
+
+	p.PacketType = 0xAE2;
+	switch (UIType) {
+	case 5: // receive 5 for ATTENDANCE_UI

Your enum ui_types seems to be not used and does not match with your code
here.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/HerculesWS/Hercules/pull/1990#pullrequestreview-102901092>,
or mute the thread

**Comment by 4144** [comment]
If this pr complete, probably better remove WIP from name

**Comment by OmarAcero** [comment]
This issues because is it caused?
Note:  **[Here](https://imgur.com/a/WCyN3)**

I am using: 2018-03-21aRagexeRE

**Comment by Asheraf** [comment]
@OmarAcero you need to enable this https://github.com/HerculesWS/Hercules/blob/stable/src/common/mmo.h#L71 and rebuild

File: src/map/clif.h [hunk 8]
```diff
@@ -574,6 +576,19 @@ enum CZ_CONFIG {
 	CZ_CONFIG_PET_AUTOFEEDING        = 2,
 	CZ_CONFIG_HOMUNCULUS_AUTOFEEDING = 3,
 };
+/**
+* Client UI types
+**/
+enum ui_types {
```
Comment (4144) [inline src/map/clif.h]: this enum probably better use in function where it belong
i mean replace ``int32 UIType`` to ``enum ui_types UIType``

**Comment by OmarAcero** [comment]
This system will be implemented in the master, or is in observation because were found problems?

File: src/map/clif.c [hunk 9]
```diff
@@ -20411,6 +20411,194 @@ void clif_hat_effect_single(struct block_list *bl, uint16 effectId, bool enable)
...
+}
+
+void clif_parse_open_ui_request(int fd, struct map_session_data *sd) __attribute__((nonnull(2)));
+void clif_parse_open_ui_request(int fd, struct map_session_data *sd)
+{
+
+	const struct PACKET_CZ_OPEN_UI *p = RP2PTR(fd);
+
+	if (clif->attendance_getendtime() < time(NULL)) {
+		clif->msgtable_color(sd, 3474, COLOR_RED);
```
Comment (4144) [inline src/map/clif.c]: probably here better use enum value and not int?

File: src/map/clif.c [hunk 10]
```diff
@@ -20411,6 +20411,194 @@ void clif_hat_effect_single(struct block_list *bl, uint16 effectId, bool enable)
...
+{
+#if PACKETVER_RE_NUM >= 20180307 || PACKETVER_MAIN_NUM >= 20180404
+
+	struct rodex_message msg = { 0 };
+	struct attendance_entry *entry;
+	int attendance_count;
+	char title[RODEX_TITLE_LENGTH], body[MAIL_BODY_LENGTH];
+
+	if (clif->attendance_getendtime() < time(NULL)) {
+		clif->msgtable_color(sd, 3474, COLOR_RED);
```
Comment (4144) [inline src/map/clif.c]: and here

**Pull request merged**